apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php

```diff
@@ -0,0 +1,100 @@
+<?php
+/**
+ * Feature flags file.
+ *
+ * @package full-site-editing
+ */
+
+/**
+ * Class A8C_Full_Site_Editing_Feature_Flags
+ */
+class A8C_Full_Site_Editing_Feature_Flags {
+	/**
+	 * Class instance.
+	 *
+	 * @var A8C_Full_Site_Editing_Feature_Flags
+	 */
+	private static $instance = null;
+
+	/**
+	 * Feature flags.
+	 *
+	 * @var array
+	 */
+	private $flags = array();
+
+	/**
+	 * A8C_Full_Site_Editing_Feature_Flags constructor.
+	 */
+	private function __construct() {
+		$this->set_flags();
+	}
+
+	/**
+	 * Creates instance.
+	 *
+	 * @return \A8C_Full_Site_Editing_Feature_Flags
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Set the feature flags.
+	 */
+	private function set_flags() {
+		$flags = '';
+
+		if ( isset( $_GET[ 'fse_flags' ] ) ) {
+			if ( empty( $_GET[ 'fse_flags' ] ) ) {
+				setcookie( 'fse_flags', '', time() - 3600 );
+				$this->flags = array();
+				return;
+			}
+
+			setcookie( 'fse_flags', $_GET[ 'fse_flags' ] );
+			$flags = $_GET[ 'fse_flags' ];
+		} else if( isset( $_COOKIE[ 'fse_flags' ] ) ) {
+			$flags = $_COOKIE[ 'fse_flags' ];
+		}
```

Comment:
let's watch our nesting as it can quickly become hard to follow

I like to use helper bindings/variables to make the repeated logic less noisy

```php
$flags = null;

$has_flags_arg    = isset( $_GET['fse_flags'] );
$has_flags_cookie = isset( $_COOKIE['fse_flags'] );

if ( $has_flags_arg && empty( $_GET['fse_flags'] ) ) {
	setcookie( 'fse_flags', '', time() - 3600 );
	$this->flags = array();
	return;
}

if ( $has_flags_arg ) {
	setcookie( 'fse_flags', $_GET['fse_flags'] );
	$flags = $_GET['fse_flags'];
} else if ( $has_flags_cookie ) {
	$flags = $_COOKIE['fse_flags'];
}
```

if this all ends up a bit confusing we can consider moving it into a helper method
```php
// no side-effects here
private function get_requested_flags() {
	if ( isset( $_GET['fse_flags'] ) && ! empty( $_GET['fse_flags'] ) ) {
		return $_GET['fse_flags'];
	}

	if ( isset( $_COOKIE['fse_flags'] ) ) {
		return $_COOKIE['fse_flags'];
	}

	return '';
}

// all side-effects here
private function set_flags() {
	$flags = $this->get_requested_flags();

	// update cookie with new information - this may differ
	// from the previous flags from our cookie and may be
	// removing flags that were present
	$cookie_expiration = empty( $flags ) ? time() - 3600 : 0;
	setcookie( 'fse_flags', $flags, $cookie_expiration );

	if ( empty( $flags ) ) {
		$this->flags = array();
		return;
	}

	foreach ( explode( ',', $flags ) as $flag ) {
		$is_removing = '-' === $flag[0];
		$name = $is_removing ? substr( $flag, 1 ) : $flag;

		$this->flags[ $name ] = ! $is_removing;
	}
}
```

Reply:
Refactored according to the first suggestion in 011120a668d1d0887a6a013658f1c06b012a69d5.

The second one is slightly different in terms of functionality - it's not handling the case of removing all of the flags when empty parameter value is passed correctly. We'd have to split that check out of `get_requested_flags`, so probably not worth it.